File: packages/ios/Sources/Components/Chip/Chip.swift
import SwiftUI

public enum KozmosChipVariant {
    case neutral
    case brand
    case destructive
}

public enum KozmosChipSize {
    case sm
    case `default`
    case lg
}

public struct KozmosChip<Icon: View>: View {
    public let text: String
    public let variant: KozmosChipVariant
    public let size: KozmosChipSize
    public let selected: Bool
    public let disabled: Bool
    public let icon: Icon?
    public let onRemove: (() -> Void)?
    public let action: (() -> Void)?

    public init(
        text: String,
        variant: KozmosChipVariant = .neutral,
        size: KozmosChipSize = .default,
        selected: Bool = false,
        active: Bool? = nil,
        disabled: Bool = false,
        onRemove: (() -> Void)? = nil,
        @ViewBuilder icon: () -> Icon,
        action: (() -> Void)? = nil
    ) {
        self.text = text
        self.variant = variant
        self.size = size
        self.selected = active ?? selected
        self.disabled = disabled
        self.icon = icon()
        self.onRemove = onRemove
        self.action = action
    }

    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
            if let icon {
                icon
                    .frame(width: 16, height: 16)
            }

            Text(text)
                .font(.system(size: fontSize, weight: .medium))
                .lineLimit(1)

            if let onRemove {
                Button(action: onRemove) {
                    Image(systemName: "xmark")
                        .font(.system(size: 10, weight: .bold))
                        .frame(width: 20, height: 20)
                        .contentShape(Circle())
                }
                .buttonStyle(.plain)
                .disabled(disabled)
                .accessibilityLabel("Remove \(text)")
            }
        }
        .padding(.horizontal, horizontalPadding)
        .padding(.vertical, verticalPadding)
        .frame(minHeight: minHeight)
        .foregroundColor(foregroundColor)
        .background(backgroundColor)
        .clipShape(Capsule())
        .overlay(
            Capsule().stroke(borderColor, lineWidth: 1)
        )
        .opacity(disabled ? 0.5 : 1)
        .contentShape(Capsule())
        .onTapGesture {
            guard !disabled else { return }
            action?()
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(text)
        .accessibilityAddTraits(action == nil ? [] : .isButton)
    }

    private var horizontalPadding: CGFloat {
        switch size {
        case .sm:
            return KozmosDimensions.primitivesLayoutSpacing100
        case .default:
            return KozmosDimensions.primitivesLayoutSpacing150
        case .lg:
            return KozmosDimensions.primitivesLayoutSpacing200
        }
    }

    private var verticalPadding: CGFloat {
        return KozmosDimensions.primitivesLayoutSpacing0
    }

    private var fontSize: CGFloat {
        switch size {
        case .sm:
            return 12
        case .default, .lg:
            return 14
        }
    }

    private var minHeight: CGFloat {
        switch size {
        case .sm:
            return 28
        case .default:
            return 32
        case .lg:
            return 36
        }
    }

    private var backgroundColor: Color {
        if selected {
            switch variant {
            case .destructive:
                return KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
            case .neutral, .brand:
                return KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
            }
        }

        switch variant {
        case .neutral:
            return KozmosColors.primitivesColorsBackground0
        case .brand:
            return KozmosColors.primitivesColorsTheme0
        case .destructive:
            return KozmosColors.primitivesColorsEmotionalDanger0
        }
    }

    private var foregroundColor: Color {
        if selected {
            switch variant {
            case .destructive:
                return KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle
            case .neutral, .brand:
                return KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
            }
        }

        switch variant {
        case .neutral:
            return KozmosColors.primitivesColorsForeground100
        case .brand:
            return KozmosColors.componentsSecondaryButtonsThemedButtonForegroundContentIdle
        case .destructive:
            return KozmosColors.componentsSecondaryButtonsDangerButtonForegroundContentIdle
        }
    }

    private var borderColor: Color {
        if selected {
            switch variant {
            case .destructive:
                return KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
            case .neutral, .brand:
                return KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
            }
        }

        switch variant {
        case .neutral:
            return KozmosColors.primitivesColorsBackground200
        case .brand:
            return KozmosColors.primitivesColorsTheme200
        case .destructive:
            return KozmosColors.primitivesColorsEmotionalDanger200
        }
    }
}

public extension KozmosChip where Icon == EmptyView {
    init(
        text: String,
        variant: KozmosChipVariant = .neutral,
        size: KozmosChipSize = .default,
        selected: Bool = false,
        active: Bool? = nil,
        disabled: Bool = false,
        onRemove: (() -> Void)? = nil,
        action: (() -> Void)? = nil
    ) {
        self.text = text
        self.variant = variant
        self.size = size
        self.selected = active ?? selected
        self.disabled = disabled
        self.icon = nil
        self.onRemove = onRemove
        self.action = action
    }
}

public struct KozmosChipGroup<Content: View>: View {
    let spacing: CGFloat
    let content: Content

    public init(
        spacing: CGFloat = KozmosDimensions.primitivesLayoutSpacing100,
        @ViewBuilder content: () -> Content
    ) {
        self.spacing = spacing
        self.content = content()
    }

    public var body: some View {
        if #available(iOS 16.0, macOS 13.0, *) {
            ViewThatFits(in: .horizontal) {
                HStack(spacing: spacing) { content }
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: spacing) { content }
                }
            }
        } else {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: spacing) { content }
            }
        }
    }
}

File: packages/android/src/main/java/com/kozmos/components/Chip/Chip.kt
package com.kozmos.components.chip

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions

enum class ChipVariant {
    Neutral, Brand, Destructive
}

enum class ChipSize {
    Sm, Default, Lg
}

@Composable
fun KozmosChip(
    text: String,
    modifier: Modifier = Modifier,
    variant: ChipVariant = ChipVariant.Neutral,
    size: ChipSize = ChipSize.Default,
    selected: Boolean = false,
    active: Boolean? = null,
    enabled: Boolean = true,
    leadingIcon: (@Composable () -> Unit)? = null,
    onRemove: (() -> Unit)? = null,
    onClick: (() -> Unit)? = null
) {
    val isSelected = active ?: selected
    val colors = chipColors(variant = variant, selected = isSelected)
    val metrics = chipMetrics(size)

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(percent = 50),
        color = colors.container,
        contentColor = colors.content,
        border = BorderStroke(1.dp, colors.border)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .heightIn(min = metrics.minHeight)
                .then(
                    if (onClick != null) {
                        Modifier.clickable(enabled = enabled) { onClick() }
                    } else {
                        Modifier
                    }
                )
                .padding(
                    start = metrics.horizontalPadding,
                    end = if (onRemove != null) KozmosDimensions.primitivesLayoutSpacing50 else metrics.horizontalPadding,
                    top = metrics.verticalPadding,
                    bottom = metrics.verticalPadding
                )
        ) {
            if (leadingIcon != null) {
                Box(
                    modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing200),
                    contentAlignment = Alignment.Center
                ) {
                    leadingIcon()
                }
                Spacer(modifier = Modifier.width(KozmosDimensions.primitivesLayoutSpacing50))
            }

            Text(
                text = text,
                color = colors.content,
                fontSize = metrics.fontSize,
                fontWeight = FontWeight.Medium,
                maxLines = 1
            )

            if (onRemove != null) {
                Spacer(modifier = Modifier.width(KozmosDimensions.primitivesLayoutSpacing50))
                Box(
                    modifier = Modifier
                        .size(KozmosDimensions.primitivesLayoutSizing300)
                        .clip(CircleShape)
                        .clickable(enabled = enabled) { onRemove() },
                    contentAlignment = Alignment.Center
                ) {
                    Icon(
                        imageVector = Icons.Default.Close,
                        contentDescription = "Remove $text",
                        modifier = Modifier.size(12.dp),
                        tint = colors.content
                    )
                }
            }
        }
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
fun KozmosChipGroup(
    modifier: Modifier = Modifier,
    horizontalSpacing: Dp = KozmosDimensions.primitivesLayoutSpacing100,
    verticalSpacing: Dp = KozmosDimensions.primitivesLayoutSpacing100,
    content: @Composable () -> Unit
) {
    FlowRow(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(horizontalSpacing),
        verticalArrangement = Arrangement.spacedBy(verticalSpacing)
    ) {
        content()
    }
}

private data class ChipColors(
    val container: Color,
    val content: Color,
    val border: Color
)

private data class ChipMetrics(
    val horizontalPadding: Dp,
    val verticalPadding: Dp,
    val minHeight: Dp,
    val fontSize: androidx.compose.ui.unit.TextUnit
)

private fun chipColors(variant: ChipVariant, selected: Boolean): ChipColors {
    if (selected) {
        return when (variant) {
            ChipVariant.Destructive -> ChipColors(
                container = KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle,
                content = KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle,
                border = KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
            )
            ChipVariant.Neutral,
            ChipVariant.Brand -> ChipColors(
                container = KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle,
                content = KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle,
                border = KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
            )
        }
    }

    return when (variant) {
        ChipVariant.Neutral -> ChipColors(
            container = KozmosColors.primitivesColorsBackground0,
            content = KozmosColors.primitivesColorsForeground100,
            border = KozmosColors.primitivesColorsBackground200
        )
        ChipVariant.Brand -> ChipColors(
            container = KozmosColors.primitivesColorsTheme0,
            content = KozmosColors.componentsSecondaryButtonsThemedButtonForegroundContentIdle,
            border = KozmosColors.primitivesColorsTheme200
        )
        ChipVariant.Destructive -> ChipColors(
            container = KozmosColors.primitivesColorsEmotionalDanger0,
            content = KozmosColors.componentsSecondaryButtonsDangerButtonForegroundContentIdle,
            border = KozmosColors.primitivesColorsEmotionalDanger200
        )
    }
}

private fun chipMetrics(size: ChipSize): ChipMetrics = when (size) {
    ChipSize.Sm -> ChipMetrics(
        horizontalPadding = KozmosDimensions.primitivesLayoutSpacing100,
        verticalPadding = 0.dp,
        minHeight = 28.dp,
        fontSize = 12.sp
    )
    ChipSize.Default -> ChipMetrics(
        horizontalPadding = KozmosDimensions.primitivesLayoutSpacing150,
        verticalPadding = 0.dp,
        minHeight = 32.dp,
        fontSize = 14.sp
    )
    ChipSize.Lg -> ChipMetrics(
        horizontalPadding = KozmosDimensions.primitivesLayoutSpacing200,
        verticalPadding = 0.dp,
        minHeight = 36.dp,
        fontSize = 14.sp
    )
}
